### PR TITLE
Android build updates

### PR DESCRIPTION
**Pull request by christianmemije**
## Summary

* Updates android build. Depends on https://github.com/learningequality/kolibri-kivy/pull/15

### CODE REVIEW

**Comment by benjaoming** [comment]
Hi @christianmemije !

Some questions:

1. Why does the Android need its own source dist versions? Are you planning to distribute it outside of the release flow of the normal source dist?
1. Are the current version numbering schemes not adequate? Why not?
1. Would this be valuable for other installers, and should it be generalized?

**Comment by jamalex** [comment]
Short answer: Python-for-Android converts a numeric period-separated version into the "pure integer" version that Android (and the Play Store) need. This resulting number needs to be monotonically increasing otherwise you can't install on top of an older version, and we want to be able to upgrade within alpha/beta versions, for testing. The scheme here was the best way we could think of to accomplish that. Don't see any usefulness outside of Android.

File: kolibri/utils/version.py [hunk 1]
```diff
@@ -412,3 +412,36 @@ def get_version(version=None):
...
+def get_android_version(version=None):
+    """
+    Returns a version to be used to build the android apk.
+    Format: [major].[minor].[commit_count]
+    Example: 0.6.1700
+    """
+    version = get_complete_version(version)
+    major_and_minor = '.'.join([str(version[0]), str(version[1])])
+    commit_count = get_git_commit_count()
+    android_version = '.'.join([major_and_minor, str(commit_count)])
```
Comment (jamalex) [inline kolibri/utils/version.py]: These three lines can be collapsed into a single `join`

**Comment by christianmemije** [comment]
Not ready for merge. Looking into https://github.com/learningequality/kolibri/pull/2199#issuecomment-330227553

**Comment by benjaoming** [comment]
> Short answer: Python-for-Android converts a numeric period-separated version into the "pure integer" version that Android (and the Play Store) need. This resulting number needs to be monotonically increasing otherwise you can't install on top of an older version

Thanks for explaining @jamalex 

@christianmemije - the version scheme as described by @jamalex is ideally property of the installer build and not the source distribution. For instance, you may change a build but base it on identical source dist versions. You could also say that we have separate release flows: 1) For the source dist 2) For the Android installer (based on the source dist)

Thus, basing this on the number of git commits as seen by the source dist wouldn't do all the work.

An example is the filenames of our Debian packages: `kolibri-0.6.0-0ubuntu0.deb`. Essentially, we add information about the build number to the versioning, because we may change things in the installer, but still base it on the same source release.

I see that we might be able to explore this to achieve a unique versioning scheme for the Android release without touching our own source code base.....

http://python-for-android.readthedocs.io/en/latest/distutils/

> All of these automatic arguments can be overridden by passing them manually on the command line, e.g.:
> 
>     python setup.py apk --name="Testapp Setup" --version=2.5

So I was thinking something like for the build command, then we don't get installer-specific stuff in the source dist:

>     kolibri_major_version=`python -c "import kolibri; print('.'.join(map(str, kolibri.VERSION[:3])))"`
>     kolibri_commit_count=`cd kolibri; git rev-list --count HEAD`
>     android_build_commit_count=`git rev-list --count HEAD`
>     android_version="$kolibri_version.$kolibri_commit_count.$android_build_commit_count"
>     
>     python kolibri/setup.py apk --name="Testapp Setup" --version=$android_version

**Comment by indirectlylit** [comment]
Added [some thoughts on installer versioning](https://learningequality.slack.com/archives/C0LE9NLCE/p1505882504000037) to Slack so as not to derail resolution of this issue.

> we may change things in the installer, but still base it on the same source release

That is a real and important case, but it's also an edge case. We have a short-term need to get Kolibri building APKs again, so if that's the only blocker I'd recommend merging and creating a new tech-debt issue to address this.

**Comment by benjaoming** [comment]
@indirectlylit I tried to suggest the main components of a none-source-impact solution above which doesn't inherit the problem of not having a build-version component.. I humbly suggest @christianmemije could try it out and see if it's possible to get around this PR and any consequential issues :)

**Comment by DXCanas** [comment]
@benjaoming 

> So I was thinking something like for the build command, then we don't get installer-specific stuff in the source dist:

The solution you described seems to take place in `./buildkite/build_apk.sh`. If that's the case, which I assume it is, since that's the only place the directories can be reliably assumed, we might into issues with passing in arguments that involve the counts of both the `kolibri-kivy` and `kolibri` repos.

You're right about `python-for-android` allowing for custom arguments in the build command --
 we're aren't using straight `python-for-android`, though. We use [`buildozer'](https://github.com/kivy/buildozer) to construct the command for us. 

We can override any value that `buildozer.spec` provides `python-for-android` by creating an environment variable, however `build_apk.sh` relies on a docker image to create the environment in which `kolibri-kivy` is set up to run. Here's the build command in [`Dockerfile`](https://github.com/learningequality/kolibri-kivy/blob/master/Dockerfile):

```
# Build the APK
RUN buildozer android debug
```

 but I imagine the python instance being wrapped in Docker would complicate the passing down of  variables based off of 2 separate repos down to the environment it's being run?

I think it's possible, but as @indirectlylit mentioned, it might be best to put off a "proper" solution until we're a) more certain of the approach we want to take and b) we're not in crunch time to have up-to-date APK's up and running. The issue has already cost me some time in #2068

**Comment by DXCanas** [comment]
A quick alternative I see is if we reduce the logic to be a simple transformation of Kolibri's `VERSION`, then move the logic from `kolibri/utils/version.py` to the `kolibri-kivy` repo's `build.sh`.

We can pass the creation of the environment variable to docker in `build.sh' then stop defining it in `buildozer.spec`. Because the android-custom version transformation only needs information about the Kolibri version, which it has inside `/code`, it should be relatively quick and feasible?

**Comment by christianmemije** [comment]
Do not merge.

**Comment by benjaoming** [comment]
Thanks @DXCanas for sharing all those insights. I'm kind of wondering if we are going for the wrong solution knowing that a better and possibly simple one exists?

Would you guys want that I try to implement my suggestion from above? I'm supposing this is all Docker-based and can be dev'ed on Ubuntu? I haven't work with Buildozer before at all.

**Comment by DXCanas** [comment]
@christianmemije is actually working on it! He and I brainstormed a bit yesterday and took a lot of what you said into account, but he'll have the most up to date news.

**Comment by christianmemije** [comment]
@benjaoming Yes currently following the suggestions and moving the android app versioning logic to the android repo. Working locally, just need to to get the docker build to work.

**Comment by christianmemije** [comment]
@benjaoming https://github.com/learningequality/kolibri-kivy/pull/14

**Comment by christianmemije** [comment]
Apk was successfully built. Merging.

**Pull request merged**